### ISSUE TITLE
Workflows permissions [epic?]: define new roles, states, transitions and permissions.

### ISSUE THREAD

**Issue by finnlewis** [localgovdrupal/localgov_core#92]
## User story

**As a:** product owner
**I want:** publishing workflow roles and permissions to be enabled by default when I enable to localgov_workflows
**So that:** we have a sensible starting point for publishing workflows.

## Notes: 

These can always be changed once installed, this is meant as a starting point for councils that want to delegate authoring permissions to a wider set of users while retaining review and publishing permisisons to a few.

## Acceptance tests: 

See 
 - [Miro board](https://miro.com/app/board/o9J_lHm9M2s=/)
 - [Permissions matrix](https://docs.google.com/spreadsheets/d/1EXkhl0K1YRpbWxbPDMQOudT9Z6CSpf9dpXsl9XJwHXs/edit#gid=0)

From these documents, we have the following: 

Contributor [contributor]:
- can write content (including creating paragraph components, upload media etc)
- reuse paragraphs from the paragraph library
- cannot publish
- sends content to approval by an Editor
- typically a service team member

Author [author]:

- has elevated access rights
- can write content
- can publish their own content
- cannot publish content sent for approval
- typically a digitally savvy service team member or someone in comms
- can (optionally) send their contents for peer review

Editor [editor]:

- can write and publish content
- can approve or reject edits
- access to an approvals dashboard
- typically members of central digital team
- can edit menus, headers, footers (and other areas which don't have workflow)

We also noted that we need a user manager. 

User manager: 

- can add users
- can assign certain roles (contributor, author, editor) to users (but not assign the super user / administrator role)

## Notes

LocalGov Drupal with the localgov_workflows module currently defines the following roles: 

- Content designer (machine name: editor)
- Author

We will need to decide if the permissions will be kept in the [localgov_roles module](https://github.com/localgovdrupal/localgov_core/tree/2.x/modules/localgov_roles) or [localgov_workflows](https://github.com/localgovdrupal/localgov_core/tree/2.x/modules/localgov_workflows), or in specific feature modules. 

## Tasks

- [x] Add new roles 
- [x] Ensure correct workflow states are in place
- [x] Configure permissions
- [x] Write phpunit tests for each role
- [x] Enable translations to test revisions / workflows compatibility ( not planning to ship with this, just to make sure we catch any architectural issues. )
- [x] Ensure the default workflows are enabled on all content types. 
- [ ] Install permissions in modules for all content types.

**Comment by stephen-cox** [comment]
@msayoung I've a question on the author permissions. Should authors be able to edit and publish other users content?

If authors can edit other people's content, then they can also do the same operations on other content as their own. This means they can move content straight from draft to published for their own and other people's content. Is this going to be a problem? If it is I'll have to look at implementing some permissions so that the authors have different permissions depending on whether they are editing their content or someone else's.

I hope that makes sense, let me know if it's not clear.

**Comment by msayoung** [comment]
Hey @stephen-cox
- my understanding was that authors can edit and publish their own, but not other people's.
- I hadn't considered authors editing (but not publishing) other people's content, i'll take that to the group

**Comment by stephen-cox** [comment]
Content Workflow Working Group clarified that for now authors shouldn't be able to edit other people's content, which solves the issue of them being able to publish other people's content: they can't.

**Comment by stephen-cox** [comment]
@willguv @msayoung This work has been deployed to the test site: https://test.localgovdrupal.org/

I have created users with each of the redefined roles: LocalGov Editor, LocalGov Author, LocalGov Contributor, and LocalGov User Manager.

To aid testing I have installed the Masquerade module which will allow you to easily switch between each of these users while testing.

If you login as the LocalGov Admin user (password pinned in the Slack #technical-group channel), you'll see the masquerade block in the footer so you can switch users.

**Comment by willguv** [comment]
Thanks @stephen-cox, looking forward to trying it

This is an invite to the other working group members too (let me know if you need log in details)

@OliverHannan 
David Siddall - looking for his GitHub name
@tom-steel

**Comment by OliverHannan** [comment]
Hello @willguv, yes i need log in details

Nice one @stephen-cox thanks so much

@davidsiddall - Your work in action!

**Comment by davidsiddall** [comment]
Nice one @stephen-cox.

I will also need log in details.

**Comment by msayoung** [comment]
Hi @stephen-cox 

I've tested the 4 roles and it seems to work as suggested, though currently limited to the 3 main Service content types

A couple of minor things
-  All of the new roles get an Access Denied when trying to add a new component on a Services Page
- Localgov Author and Localgov Editor cannot Archive content (it's not mentioned above but it's on the spreadsheet


Notes for discussion:
- Content Editor cannot edit Localgov Admin's content (is this latter role going to remain?)
- Because we can't reference unpublished content for a Service parent / child - there may be another step *after* content has been reviewed and published to go and update those links. (some of those pieces of content may have been created by others and therefore can't be edited)

**Comment by stephen-cox** [comment]
Thanks for testing this @msayoung 

> All of the new roles get an Access Denied when trying to add a new component on a Services Page

I needed to give the roles permission to use the entity browser. Should work now.

> Localgov Author and Localgov Editor cannot Archive content (it's not mentioned above but it's on the spreadsheet

Editors should be able to archive content (that is, move draft or published content to archived)
I have given this permission to authors.

> Content Editor cannot edit Localgov Admin's content (is this latter role going to remain?)

Editors should be able to edit any content. We haven't implemented the permissions for this for anything other than service landing, sublanding and pages yet. These permissions will need to be assigned on each module and I'll wait until the role naming question is resolved before creating PRs for all of these.

LocalGov Admin is not a role but the name of the Drupal user 1 account on the test and demo sites.

> we can't reference unpublished content for a Service parent / child

Yep - a tricky problem. There _may_ be a workable solution to this so that draft pages can be previewed while referencing other draft pages, but will need some investigation

**Comment by msayoung** [comment]
Thanks @stephen-cox , that's great and thanks for the clarification. 

I'll seek some more feedback from the others.

**Comment by OliverHannan** [comment]
> All of the new roles get an Access Denied when trying to add a new component on a Services Page

This worked for me on all new users

> Localgov Author and Localgov Editor cannot Archive content 

I was able to archive as Author but **not** as Editor, see image - option wasn't available to archive 

![cant archive](https://user-images.githubusercontent.com/47107641/124960613-6695e780-e014-11eb-8427-a90a32a84e2a.png)

> Content Editor cannot edit Localgov Admin's content (is this latter role going to remain?)

@stephen-cox thanks for explaination - makes sense

> we can't reference unpublished content for a Service parent / child

@stephen-cox again, makes sense. Cheers

**Comment by msayoung** [comment]
In the review we discovered that it's not possible to Archive content when it's in Review. 
@stephen-cox will update this. 
Stephen will also add these permissions to the other content types. 

We discussed the following and decided that we can discuss and raise separate issues if they are determined as necessary
- roles will go ahead as originally defined, keeping in mind that the labels can be changed 
- Authors may need to edit other people's content. This is a bit more complex as we have to work against Drupal's normal permissions. I'll raise a ticket for this. 
- Changing the label of Drupal's "Authored by" field
